llvm/lib/Target/SystemZ/SystemZISelLowering.cpp

```diff
@@ -8701,95 +8734,341 @@ SDValue SystemZTargetLowering::combineSETCC(
+      return {};
+    // Avoid introducing CC spills (because SRA would clobber CC).
+    if (!N->hasOneUse())
+      return {};
+    SDValue IPM = SHL->getOperand(0);
+    const auto &&SDVals = simplifyAssumingCCVal(IPM, CC, DCI);
+    if (SDVals.empty())
+      return SDVals;
+    auto SRAShift = SRACount->getZExtValue();
+    auto SHLShift = SHLCount->getZExtValue();
+    SmallVector<SDValue, 4> ShiftedVals;
+    for (const auto &SDVal : SDVals) {
+      SDValue SRAVal = DAG.getNode(ISD::SHL, DL, MVT::i32, SDVal,
+                                   DAG.getConstant(SHLShift, DL, MVT::i32));
+      ShiftedVals.emplace_back(
+          DAG.getNode(ISD::SRA, DL, MVT::i32, SRAVal,
+                      DAG.getConstant(SRAShift, DL, MVT::i32)));
+    }
+    return ShiftedVals;
+  }
+  if (N->getOpcode() == SystemZISD::SELECT_CCMASK) {
+    SDValue TrueVal = N->getOperand(0), FalseVal = N->getOperand(1);
+    auto *TrueOp = TrueVal.getNode();
+    auto *FalseOp = FalseVal.getNode();
+    auto *CCValid = dyn_cast<ConstantSDNode>(N->getOperand(2));
+    auto *CCMask = dyn_cast<ConstantSDNode>(N->getOperand(3));
+    if (!TrueOp || !FalseOp || !CCValid || !CCMask)
+      return {};
+
+    int CCValidVal = CCValid->getZExtValue();
+    int CCMaskVal = CCMask->getZExtValue();
+    const auto &&TrueSDVals = simplifyAssumingCCVal(TrueVal, CC, DCI);
+    const auto &&FalseSDVals = simplifyAssumingCCVal(FalseVal, CC, DCI);
+    if (TrueSDVals.empty() || FalseSDVals.empty())
+      return {};
+    SDValue Op4CCReg = N->getOperand(4);
+    auto *Op4CCNode = Op4CCReg.getNode();
+    if (Op4CCNode && Op4CCNode != CCNode)
+      combineCCMask(Op4CCReg, CCValidVal, CCMaskVal, DCI);
+    Op4CCNode = Op4CCReg.getNode();
+    if (!Op4CCNode || Op4CCNode != CCNode)
+      return {};
+    SmallVector<SDValue, 4> MergedSDVals;
+    for (auto &CCVal : {0, 1, 2, 3})
+      MergedSDVals.emplace_back((((CCMaskVal & (1 << (3 - CCVal))) != 0) &&
+                                 ((CCValidVal & (1 << (3 - CCVal))) != 0))
+                                    ? TrueSDVals[CCVal]
+                                    : FalseSDVals[CCVal]);
+    return MergedSDVals;
+  }
+  SDValue Op0, Op1;
+  unsigned Opcode;
+  if (isValidBinaryOperation(Val, Op0, Op1, Opcode)) {
+    const auto &&Op0SDVals = simplifyAssumingCCVal(Op0, CC, DCI);
+    const auto &&Op1SDVals = simplifyAssumingCCVal(Op1, CC, DCI);
+    if (Op0SDVals.empty() || Op1SDVals.empty())
+      return {};
+    SmallVector<SDValue, 4> BinaryOpSDVals;
+    for (auto CCVal : {0, 1, 2, 3})
+      BinaryOpSDVals.emplace_back(DAG.getNode(
+          Opcode, DL, Val.getValueType(), Op0SDVals[CCVal], Op1SDVals[CCVal]));
+    return BinaryOpSDVals;
+  }
+  return {};
+}
+
+bool SystemZTargetLowering::combineCCMask(SDValue &CCReg, int &CCValid,
+                                          int &CCMask,
+                                          DAGCombinerInfo &DCI) const {
```

Comment:
Looks like these only really need the DAG not the DCI?   And if we do pass in the DAG, is there another reason to make them member functions?